### PR TITLE
Playwright script to take screenshot of the login page

### PR DESCRIPTION
**Pull request by niranjan-uma-shankar**
#### Changes proposed in this Pull Request

For context, check pau2Xa-3my-p2.

* Playwright script that takes a full page screenshot of login, signup, and checkout pages. 
* The screenshots are uploaded to a custom WP.com REST API, which in turn uploads them to a WP.com blog. 

#### Testing instructions

Depends on D69547-code.

* Follow the instructions in p4TIVU-8Qw-p2 to setup e2e tests locally
* Checkout this branch in your local Calypso instance
* In test/e2e folder of your local Calypso folder, run `yarn install`
* Apply D69547-code and follow the setup instructions written in that patch. 
* Run the Playwright script using `BABEL_ENV=test yarn jest specs/specs-playwright/wp-signup__screenshot.ts`.
* Verify that fullpage screenshots of signup, login, and checkout are stored on the blog, and the ToS link is visible in all these pages.

Related to #

### CODE REVIEW

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+			await fetch( 'https://public-api.wordpress.com/wpcom/v2/screenshots', {
+				method: 'POST',
+				body: form,
+			} );
+			await loginPage.signup();
+		} );
+
+		it( 'Sign up as new user', async function () {
+			const userSignupPage = new UserSignupPage( page );
+			await page.screenshot( { path: 'signup.png' } );
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: Just a heads up, initializing a new page object (like here) doesn't actually do anything unless you call methods from it (eg. `userSignupPage.signup`).

If you have any feedback on the way methods are named in the page objects, or how the library is structured, I'd love to hear some feedback - I am realizing now that some method names in `LoginPage` are a not quite as descriptive as I had initially thought, for example.

Comment (niranjan-uma-shankar) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: I'll take a closer look and share feedback, but top of my mind it's looking okay to me.

**Comment by scinos** [comment]
This should be in a separate directory or package, not in `test/e2e` (as it is not testing anything).

File: packages/calypso-e2e/src/lib/components/navbar-component.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ const selectors = {
 	writeButton: '*css=a >> text=Write',
 	notificationsButton: 'a[href="/notifications"]',
 	meButton: 'a[data-tip-target="me"]',
+	closeCheckout: 'a[data-tip-target="close"]',
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/components/navbar-component.ts]: This method should be implemented in https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts instead, since the navbar is strictly only for items on the always-showing navbar.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 3]
```diff
@@ -31,32 +31,85 @@ describe( DataHelper.createSuiteTitle( 'Signup: WordPress.com Free' ), function
...
+			let form = new FormData();
 			form.append( 'my_file', fs.createReadStream( 'login.png' ) );
-			const data = { image: buffer.toString( 'base64' ) };
 			await fetch( 'https://public-api.wordpress.com/wpcom/v2/screenshots', {
 				method: 'POST',
 				body: form,
-			} );
+			} )
+			.then( response => response.json())
+			.then( response => expect( response?.upload_status ).toStrictEqual( 'success' ) );
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: Seeing that this operation is repeated once more below, I would personally put this as part of the `afterAll` hook that runs at conclusion of this file. It can perhaps read an array of files that need to be submitted to the endpoint. The array could be built up as the test steps execute, eventually ending up with an array of length 2 (in this case).

The advantage of `afterAll` hook is not only that it would eliminate duplicate code, but also it is guaranteed to run at end of the test suite. Context: with the way we've configured Jest, a failure anywhere immediately aborts the entire suite and mark unexecuted steps as `skipped`.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+			const navbarComponent = new NavbarComponent( page );
+			await navbarComponent.clickCloseCheckout();
+		} );
+	} );
+
+	describe( 'Delete user account', function () {
+		it( 'Close account', async function () {
+			const closeAccountFlow = new CloseAccountFlow( page );
+			await closeAccountFlow.closeAccount();
+		} );
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: This is more of a question - does what you are trying to accomplish require the use of a new account created at the time of the script execution?

**Comment by worldomonation** [comment]
Overall, looks like a pretty solid first pass! I like that you were able to leverage (for many portions) existing library methods.

I've made a few suggestions on top of what Sergio has suggested. 

Related to what Sergio has commented - before merging to trunk, someone from Quality Squad or Team Calypso-Platform can create a new TeamCity build configuration specifically to run this file. That way, it can be isolated in its own task and is not affected by changes to mainline E2E tests. We can do any interval you'd like, as the configuration is based on crontab-like syntax.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 5]
```diff
@@ -0,0 +1,299 @@
+/**
+ * @group calypso-release
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: This looks like a utility script and so I would like to propose that this be moved out of the pre-release tests group and instead put into a new category, such as `compliance` or `legal`.

I can help set up a new TeamCity task for this once the script is done and ready.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts [hunk 6]
```diff
@@ -75,6 +78,25 @@ export class CartCheckoutPage {
 		this.page = page;
 	}
 
+	/**
+	 * Navigates to checkout page of the specified blog
+	 *
+	 * @param blogName Blog name for which the checkout page is to be loaded.
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: nit: expected JSDoc format is `@param {string} blogName <description>`.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts [hunk 7]
```diff
@@ -75,6 +78,25 @@ export class CartCheckoutPage {
 		this.page = page;
 	}
 
+	/**
+	 * Navigates to checkout page of the specified blog
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: nit: missing full stop.

File: packages/calypso-e2e/src/lib/pages/login-page.ts [hunk 8]
```diff
@@ -76,6 +76,22 @@ export class LoginPage {
...
+		await this.page.goto( getCalypsoURL( targetUrl ), { waitUntil: 'networkidle' } );
+	}
+
+	/**
+	 * Navigates to the /log-in/{locale} endpoint.
+	 */
+	async visitWhiteLogin( locale = '' ): Promise< void > {
+		const targetUrl = locale ? `log-in/new/${ locale }` : 'log-in/new';
+		await this.page.goto( getCalypsoURL( targetUrl ), { waitUntil: 'networkidle' } );
+	}
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/login-page.ts]: It was initially not clear to me (from the JSDoc or method name) what this intended to do.

Now that I understand the intent is to navigate to one of the two variations for the login endpoint, I would like to suggest the following changes to be in line with POM design principles used elsewhere in `calypso-e2e`:

1. an optional parameter be added to the `options?: LoginOptions` interface used by `LoginPage.login`:

```
	async login(
		credentials: LoginCredentials | TestAccount,
		options?: LoginOptions
	): Promise< void > {
```

```
interface LoginOptions {
	landingUrl?: string;
    endpoint?: string;
}
```

2. specify the endpoint when calling `LoginPage.login` like:
```
await loginPage.login( { account: user }, { endpoint: 'log-in/new', landingUrl: '**/log-in/authenticator' } );
```

This way, we avoid creating single-purpose methods in POMs that can easily be integrated into the mainline methods. This helps us avoid the situation we faced with the Selenium-based framework.

File: packages/calypso-e2e/src/lib/pages/user-signup-page.ts [hunk 9]
```diff
@@ -35,6 +36,24 @@ export class UserSignupPage {
...
+
+	/**
+	 * Navigates to the /log-in endpoint.
+	 */
+	async visitWhiteSignup( locale = '' ): Promise< void > {
+		const targetUrl = locale ? `start/${ locale }` : 'start';
+		console.log( targetUrl );
+		await this.page.goto( getCalypsoURL( targetUrl ), { waitUntil: 'networkidle' } );
+	}
+
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/user-signup-page.ts]: With changes proposed in the `LoginPage` POM, changes here should no longer be necessary.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+	let page: Page;
+	let selectedDomain: string;
+
+	setupHooks( ( args ) => {
+		page = args.page;
+	} );
+
+	describe( 'Signup', function () {
+		jest.setTimeout( 1000000 );
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: FYI, changing this timeout does not change the Playwright timeout.

Generally speaking, it is preferable to leave the environment management to the test runner and framework itself.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+			'zh-cn',
+			'zh-tw',
+			'ko',
+			'ar',
+			'sv',
+		];
+
+		it( 'Screenshot blue login page in desktop viewport, en and Mag-16 locales', async function () {
+			const loginPage = new LoginPage( page );
+			for ( const locale of [ 'en', ...magnificientNonEnLocales ] ) {
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: As long as this script is not required to run as part of regular calypso deployment process, I have no issues with the `for...of` loop within a test step.

However, if this script is expected to run as part of every calypso deployment, then this loop leads to issues with TeamCity being unable to break out the individual steps taken (eg. Screenshot blue login page (en)).

In that case, I'd like to request the use of `it.each` parametrization for this step and all other steps that iterate through the locale.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 4 above)
Comment (worldomonation) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: To expand on my previous comment, is it possible to achieve the screenshot of Checkout TOS by simply logging into an existing user, add an item to the cart, navigate to the checkout page and take screenshots?

Where possible, we try to avoid creating new users or sites to help reduce burden on production databases.

**Review by worldomonation — CHANGES_REQUESTED** [review]

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 3 above)
Comment (niranjan-uma-shankar) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: I've refactored the code such that this code is run only once, by zipping all the screenshot images and making a single API request to upload the files to the backend.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 4 above)
Comment (niranjan-uma-shankar) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: Thanks for the feedback, I've removed the creation of new user.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 5 above)
Comment (niranjan-uma-shankar) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: I've changed this to a new "legal" group.

File: packages/calypso-e2e/src/lib/pages/login-page.ts (hunk 8 above)
Comment (niranjan-uma-shankar) [reply packages/calypso-e2e/src/lib/pages/login-page.ts]: As suggested, I've refactored the code and reused the existing `visitIO` function to append a path param such that the script navigates to the white background login page(`/log-in/new`) and blue background signup page(`/start/premium`)  based on the props provided to it.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 11 above)
Comment (niranjan-uma-shankar) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: I expect this script to run every time a change is made to signup, login, or checkout file paths. So, it won't run on  every Calypso deployment. So is it fine to retain the `for ... of` loop, or do you still suggest to use `it..each`?

**Comment by niranjan-uma-shankar** [comment]
Thanks for the review @worldomonation , I've pushed in commits that address your comments, and I've asked some clarifications above on some of the comments. 

> Related to what Sergio has commented - before merging to trunk, someone from Quality Squad or Team Calypso-Platform can create a new TeamCity build configuration specifically to run this file. That way, it can be isolated in its own task and is not affected by changes to mainline E2E tests. We can do any interval you'd like, as the configuration is based on crontab-like syntax.

Sounds great. I'd like this script to run every time a change to signup, login, or checkout file paths is deployed.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+				quality: 20,
+			} );
+			page.setViewportSize( { width: 1280, height: 720 } );
+		} );
+
+		it( 'Close checkout and change UI language', async function () {
+			const changeUILanguageFlow = new ChangeUILanguageFlow( page );
+			for ( const locale of magnificientNonEnLocales ) {
+				await cartCheckoutPage.clickCloseCheckout();
+				await changeUILanguageFlow.changeUILanguage( locale as LanguageSlug );
```
Comment (niranjan-uma-shankar) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts:229]: There's an issue I'm running into with the "change UI language" section which I'm not quite able to debug. The interface language switch errors out on the browser even though the correct language is selected(screenshot below). To further explain, I can see that the correct locale search term is selected in the language switcher modal, and when the "Apply" button is clicked, the language change does not get applied and the error below is seen. What makes it strange is that the change UI language works some of the times, but fails most of the times. Would you be knowing why this is happening?
cc'ing @yuliyan in case you have any ideas too.

eee

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 10 above)
Comment (niranjan-uma-shankar) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: Without this line, the script ends before finishing all the steps with the message "thrown: "Exceeded timeout of 160000 ms for a test", so that's why I added this line.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 12 above)
Comment (yuliyan) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts:229]: I suppose you're testing with sandboxed `public-api.wordpress.com`. If that's the case, could there be some changes on your sandbox, that would cause the update settings request to fail?

Could you confirm the POST request to `https://public-api.wordpress.com/rest/v1.1/me/settings/` is being sent with the expected payload - it should be something like `{"language":"de"}`. You'll need the `Preserve log` checked, as applying the language settings refreshes the page.

**Comment by scinos** [comment]
> I'd like this script to run every time a change to signup, login, or checkout file paths is deployed.

Unfortunately, that's not something our infra support. The options are run for every PR, run before every deployment to production, or run at a scheduled interval (eg: every 24h). I guess run before every deployment is our best chance here.

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 12 above)
Comment (niranjan-uma-shankar) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts:229]: I am sandboxing `public-api.wordpress.com` and I noticed that even with no changes applied to my sandbox, I see consistent failure of the change UI language step. Observing the network requests, I see the request to `/me/settings` endpoint does not include the `language` payload. Changing UI language works if `public-api.wordpress.com` is pointing to production. I don't think I understand why this is happening, but given the behaviours, I'm assuming this error won't be seen on production when this PR is deployed.

**Comment by niranjan-uma-shankar** [comment]
> Unfortunately, that's not something our infra support. The options are run for every PR, run before every deployment to production, or run at a scheduled interval (eg: every 24h). I guess run before every deployment is our best chance here.

Is it possible to setup a VCS trigger, as mentioned in p4TIVU-9Ql-p2#comment-11273, so that this runs only when certain file paths are changed?

**Comment by scinos** [comment]
> Is it possible to setup a VCS trigger, as mentioned in p4TIVU-9Ql-p2#comment-11273, so that this runs only when certain file paths are changed?

It is possible but I wouldn't recommend it, as it will potentially miss many cases (eg: if part of the flow is extracted to a separate package, translations, etc.). That's the main reason we don't use it anywhere.

Would it be possible to take a screenshot before every release (technically, after every commit to `trunk`), and delete the duplicates at the storage layer?

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts (hunk 11 above)
Comment (worldomonation) [reply test/e2e/specs/specs-playwright/wp-signup__screenshot.ts]: Generally, using Jest's built-in parallelization tool is preferred over using `for ... of` loop, so I would prefer the following:

```
it.each( ['en', ...magnificientNonEnLocales ])( 'Screenshot blue login page in desktop viewport for %s' )
```

which would generate something like

```
'Screenshot blue login page in desktop viewport for en' 
'Screenshot blue login page in desktop viewport for fr'
'Screenshot blue login page in desktop viewport for de'
```

and so on.

**Comment by worldomonation** [comment]
> > Unfortunately, that's not something our infra support. The options are run for every PR, run before every deployment to production, or run at a scheduled interval (eg: every 24h). I guess run before every deployment is our best chance here.
> 
> Is it possible to setup a VCS trigger, as mentioned in p4TIVU-9Ql-p2#comment-11273, so that this runs only when certain file paths are changed?

@niranjan-uma-shankar I would suggest that we create a new build configuration under the Calypso > Web App project and name it something like 'Terms of Service' or similar. 

The idea is to isolate this script/task to a dedicated build which can then run based on a trigger of your choosing. 
I've recently stood up end-to-end tests for P2s, and scheduled them to run every 3 hours. Would something like that work? I can get a build up and running within a day if that's what you'd like.

Once a build is up and running, we can move this script into its own standalone directory since this isn't really an end to end test for Calypso and thus should not live in `test/e2e`. Perhaps something like `wp-calypso/tools` or `wp-calypso/tos` would be better?

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts [hunk 13]
```diff
@@ -233,6 +256,15 @@ export class CartCheckoutPage {
 		await cvvInput.fill( paymentDetails.cvv );
 	}
 
+	/**
+	 * Enter cardholder's name in the payment details form.
+	 *
+	 * @param {PaymentDetails} paymentDetails Object implementing the PaymentDetails interface.
+	 */
+	async enterCardholderName( paymentDetails: PaymentDetails ): Promise< void > {
+		await this.page.fill( selectors.cardholderName, paymentDetails.cardHolder );
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: Quick question - I assume the method `CartCheckoutPage.enterPaymentDetails` is not sufficient for this?

File: packages/calypso-e2e/src/lib/pages/login-page.ts [hunk 14]
```diff
@@ -71,9 +71,12 @@ export class LoginPage {
 
 	/**
 	 * Navigates to the /log-in endpoint.
+	 *
+	 * @param {string} path The path to be appended to /log-in. E.g. /log-in/new is the reskinned login page.
 	 */
-	async visit(): Promise< void > {
-		await this.page.goto( getCalypsoURL( 'log-in' ), { waitUntil: 'networkidle' } );
+	async visit( path = '' ): Promise< void > {
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/login-page.ts]: I'd prefer that this parameter be an optional one instead of using a default value, as it makes things more explicit.

Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/login-page.ts]:
```suggestion
	async visit( {path}: {path: string}= {} ): Promise< void > {
```

File: packages/calypso-e2e/src/lib/pages/user-signup-page.ts [hunk 15]
```diff
@@ -35,6 +36,16 @@ export class UserSignupPage {
 		this.page = page;
 	}
 
+	/**
+	 * Navigates to the /start endpoint.
+	 *
+	 * @param {string} path The path to be appended to /start. E.g. /start/premium is the premium plan signup flow.
+	 */
+	async visit( path = '' ): Promise< void > {
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/user-signup-page.ts]: Same recommendation here as above.

File: packages/calypso-e2e/src/lib/pages/login-page.ts (hunk 14 above)
Comment (niranjan-uma-shankar) [reply packages/calypso-e2e/src/lib/pages/login-page.ts]: @worldomonation Not sure if the suggested syntax is correct, I'm getting an error. Could you please check the syntax I've updated in the code and let me know if that is correct?

```
async visit( path?: string ): Promise< void > {
```

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts (hunk 13 above)
Comment (niranjan-uma-shankar) [reply packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: `CartCheckoutPage.enterPaymentDetails` should be enough - I've modified the code to use this instead.

File: packages/calypso-e2e/src/lib/pages/login-page.ts (hunk 14 above)
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/login-page.ts]: > @worldomonation Not sure if the suggested syntax is correct, I'm getting an error. Could you please check the syntax I've updated in the code and let me know if that is correct?
> 
> ```
> async visit( path?: string ): Promise< void > {
> ```

I had a stray single quote in the original suggestion, which is why it didn't work.

An optional parameter is better than a default value, but I still think we're better served by explicitly passing in a named parameter. It would end up like this when it is called:

`loginPage.visit( { path: '...' } )`

instead of 

`loginPage.visit( '...' )`

Comment (niranjan-uma-shankar) [reply packages/calypso-e2e/src/lib/pages/login-page.ts]: Thanks, I've made this change. Could you please check if this looks good to you?

**Comment by niranjan-uma-shankar** [comment]
@worldomonation I've addressed all the feedback comments, could you please take another look when you get a chance? :-)

> I've recently stood up end-to-end tests for P2s, and scheduled them to run every 3 hours. Would something like that work? I can get a build up and running within a day if that's what you'd like.

That sounds good to me, we can schedule this to run once every three hours. If we can get the build up and running with this configuration, that's be great!

> Once a build is up and running, we can move this script into its own standalone directory since this isn't really an end to end test for Calypso and thus should not live in test/e2e. Perhaps something like wp-calypso/tools or wp-calypso/tos would be better?

Sure, I can move this to a standalone directory like `wp-calypso/tos`.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts [hunk 16]
```diff
@@ -13,8 +14,8 @@ const selectors = {
 	// Cart item
 	cartItem: ( itemName: string ) =>
 		`[data-testid="review-order-step--visible"] .checkout-line-item >> text=${ itemName.trim() }`,
-	removeCartItemButton: ( itemName: string ) =>
-		`[data-testid="review-order-step--visible"] button[aria-label*="Remove ${ itemName.trim() } from cart"]`,
+	removeCartItemButton: () =>
+		`[data-testid="review-order-step--visible"] button.checkout-button.checkout-line-item__remove-product`,
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: Not sure why this change is included. It does not appear to be used anywhere in rest of the PR, and it is used in `CartCheckoutPage.removeCartItem` which is in turn used by numerous specs.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts [hunk 17]
```diff
@@ -38,9 +39,11 @@ const selectors = {
 	// Tax information
 	countryCode: `select[aria-labelledby="country-selector-label"]`,
 	postalCode: `input[id="contact-postal-code"]`,
-	submitBillingInformationButton: `button[aria-label="Continue with the entered contact details"]`,
+	submitBillingInformationButton:
+		'[data-testid="contact-form--visible"] button.checkout-button.is-status-primary',
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:42]: Is this change due to I18N reasons? Otherwise, the resulting selector is not as descriptive as the original form.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts [hunk 18]
```diff
@@ -75,6 +79,21 @@ export class CartCheckoutPage {
 		this.page = page;
 	}
 
+	/**
+	 * Navigates to checkout page of the specified blog.
+	 *
+	 * @param {{path: string}: string } param1 Key/value pair of the blogname for which checkout is to be loaded.
+	 */
+	async visit( { blogName }: { blogName: string } = { blogName: '' } ): Promise< void > {
```
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: If `blogName` is optional, this would result in the following screen being shown unexpectedly:

image

I think a more robust solution would be to require the parameter `blogName`. What do you think?

**Review by worldomonation — CHANGES_REQUESTED** [review]
It's looking better and better! I have a few questions that I've left as reviews.

**Comment by worldomonation** [comment]
I'll put an item to set up the build configuration for ToS/Legal for the next sprint (coming up on 01/17).

**Comment by worldomonation** [comment]
@niranjan-uma-shankar I also suggest rebasing if possible, and re-running the Code Style build to make sure all lint issues are fixed - to prevent your changes from being reverted.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts (hunk 17 above)
Comment (niranjan-uma-shankar) [reply packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts:42]: Correct, this is due to i18n reasons. We run through the checkout page screenshot for all non-EN languages.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts (hunk 16 above)
Comment (niranjan-uma-shankar) [reply packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: Good catch, this is a remnant from an earlier change which is no longer needed. I've reverted this change.

File: packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts (hunk 18 above)
Comment (niranjan-uma-shankar) [reply packages/calypso-e2e/src/lib/pages/cart-checkout-page.ts]: Makes sense, I've refactored as you suggested.

**Comment by niranjan-uma-shankar** [comment]
>  and re-running the Code Style build to make sure all lint issues are fixed 

@worldomonation I see a few failures in Code style and a bunch of other things such as E2E Test (desktop), Unit tests (Web app) etc. It's not clear from following through the links on what's the source of the error. Could you point me to how I can learn more about the error?

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+				await domainSearchComponent.search( username + '.live' );
+				await domainSearchComponent.selectDomain( '.live' );
+
+				await page.click(
+					'[data-e2e-type="domain"].site-or-domain__choice .site-or-domain__choice-button button'
+				);
+				await page.waitForSelector( selectors.isBlueSignup );
+
+				await page.screenshot( {
+					path: `tos_blue_signup_desktop_${ locale }.png`,
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts:81]: I ran this script locally and turns out this would end up creating all screenshots at the `test/e2e` directory.

I'd suggest that in a follow-up issue this be addressed, so that all screenshots are stored in a directory, for example `test/e2e/tos-screenshots`.

**Comment by worldomonation** [comment]
Overall, looks great to me! 

I'm happy you were able to get the PR to this state where no failure of existing builds occur.

As noted in cone of my comment, I tried running this script locally and it ended up failing halfway through due to an unexpected failure to navigate. There are many other ways to optimize and make this task more resilient against failure, but let's merge this for now and iterate in subsequent PRs.

**Review by worldomonation — APPROVED** [review]

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+				form.append( 'zip_file', fs.createReadStream( zipFilename ) );
+				fetch( 'https://public-api.wordpress.com/wpcom/v2/screenshots', {
+					method: 'POST',
+					body: form,
+					headers: {
+						Authorization: `Bearer ${ bearerToken }`,
+					},
+				} )
+					.then( ( response ) => response.json() )
+					.then( ( response ) => expect( response?.upload_status ).toStrictEqual( 'success' ) );
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts:280]: For example, if this task can be performed via cURL, then it can be moved out of this script and into a build step, which would:
1. allow for multiple scripts to generate artifacts in parallel, then zip into one archive at end.
2. better control over where the artifacts end up (on TeamCity or on the MarTech servers).

File: test/e2e/specs/specs-playwright/wp-signup__screenshot.ts [hunk 21]
```diff
@@ -0,0 +1,284 @@
+/**
+ * @group legal
```
Comment (worldomonation) [inline test/e2e/specs/specs-playwright/wp-signup__screenshot.ts:2]: FYI, this file should be in `test/e2e/specs/specs-legal` or somewhere else entirely, since it's not an end-to-end test.

This can be addressed in a follow-up PR.

**Comment by niranjan-uma-shankar** [comment]
Thanks for the review @worldomonation ! I'm tracking the points you raised in https://github.com/Automattic/wp-calypso/issues/60614, https://github.com/Automattic/wp-calypso/issues/60615, and https://github.com/Automattic/wp-calypso/issues/60616.
